samza-core/src/main/java/org/apache/samza/system/inmemory/InMemorySystemFactory.java

```diff
@@ -29,22 +30,43 @@
 
 /**
  * Initial draft of in-memory {@link SystemFactory}. It is test only and not meant for production use right now.
+ *
+ * <p>Runtime configuration of InMemorySystem depends on a config "test.id"</p>
+ * <p>
+ * If {@code test.id} param is configured in the configs, it creates an isolated InMemorySystem for that test to run
+ * , all the in memory streams (input/output/intermediate) are created using this isolated InMemorySystem.
+ * </p>
+ * <p>
+ * If {@code test.id} param is not configured for an application, it shares a default InMemorySystem for all the streams
+ * This system is shared between all the applications missing {@code test.id} in their configs running in the same JVM
+ * using InMemorySystem
+ * </p>
  */
 public class InMemorySystemFactory implements SystemFactory {
-  private static final InMemoryManager MEMORY_MANAGER = new InMemoryManager();
+  private static final String TEST_ID = "test.id";
+  private static final ConcurrentHashMap<Integer, InMemoryManager> IN_MEMORY_MANAGERS = new ConcurrentHashMap<>();
+  private static final InMemoryManager DEFAULT_MANAGER = new InMemoryManager();
 
   @Override
   public SystemConsumer getConsumer(String systemName, Config config, MetricsRegistry registry) {
-    return new InMemorySystemConsumer(MEMORY_MANAGER);
+    return new InMemorySystemConsumer(getOrDefaultInMemoryManagerByTestId(config));
   }
 
   @Override
   public SystemProducer getProducer(String systemName, Config config, MetricsRegistry registry) {
-    return new InMemorySystemProducer(systemName, MEMORY_MANAGER);
+    return new InMemorySystemProducer(systemName, getOrDefaultInMemoryManagerByTestId(config));
   }
 
   @Override
   public SystemAdmin getAdmin(String systemName, Config config) {
-    return new InMemorySystemAdmin(MEMORY_MANAGER);
+    return new InMemorySystemAdmin(getOrDefaultInMemoryManagerByTestId(config));
+  }
+
+  private InMemoryManager getOrDefaultInMemoryManagerByTestId(Config config) {
+    if (!config.containsKey(TEST_ID) || config.get(TEST_ID) == null) {
```

Comment:
Maybe something like
```
String inMemoryScope = config.get(...);
if (inMemoryScope != null) {
   return IN_MEMORY_MANAGER.computeIfAbsent(inMemoryScope, new InMemoryManager());
}

return DEFAULT_MANAGER;
```

Reply:
I am also adding DEFAULT_MANAGER in ConcurrentHashMap to make it thread safe (just in case)